---
title: "Clustering"
author: "Ashley Sawle"
date: 'April 2022'
output:
  ioslides_presentation:
    widescreen: true
    smaller: true
    logo: Images/uniOfCamCrukLogos.png
    css: css/stylesheet.css
---

```{r include=FALSE}
library(tidyr)
library(dplyr)
```

## Single Cell RNAseq Analysis Workflow

```{r, echo=FALSE, out.width='70%', fig.align='center'}
knitr::include_graphics('Images/workflow2_clustering.png')
```

## Motivation

The data has been QC'd and normalized, and batch corrected.
  
We can now ask biological questions.

* unsupervised clustering: identification of groups of cells based on the
similarities of the transcriptomes without any prior knowledge of the labels
usually using the PCA output

* *de novo* discovery and annotation of cell-types based on transcription
profiles

## Single Cell RNAseq Analysis Workflow

```{r, echo=FALSE, out.width='100%', fig.align='center'}
knitr::include_graphics("../Images/Andrews2017_Fig1.png", auto_pdf = TRUE)
```

## Graph-based clustering {.columns-2 .smaller}

Nearest-Neighbour (NN) graph:

  * cells as nodes
  * their similarity as edges

In a NN graph two nodes (cells), say X and Y, are connected by an edge if:

* the distance between them is amongst the **k** smallest distances from X to
other cells, ‘**K**NN’  

or  

* the above plus the distance between them is amongst the **k** smallest
distances from X to other cells **shared**-NN (‘**S**NN).
  
Once edges have been defined, they can be weighted by various metrics.

<p class="forceBreak"></p>

```{r, include=FALSE}
require(igraph)
g <- make_graph( ~ A-C:D:F:G:H, B-D:E:F:H:I)
cols <- c("red", "red",
          "orange", "lightblue", "orange",
          "lightblue",
          "orange", "lightblue", "orange")
names(cols) <- c(LETTERS[1:9])
```

```{r, fig.width = 5, fig.height = 5, echo=FALSE}
#, fig.align="center"
plot.igraph(
  g, layout = layout_with_fr(g),
  vertex.size = 20,
  vertex.color = cols[names(V(g))]
)
```

## Graph-based clustering

```{r, echo=FALSE,out.width = '40%', fig.align="center"}
knitr::include_graphics("../Images/KNNvSNN.svg", auto_pdf = TRUE)
```

## Graph-based clustering

Example with different numbers of neighbours:

```{r, echo=FALSE, out.height='100%', out.width = '100%', fig.align="center"}
knitr::include_graphics("../Images/bioCellGenGraphDeng2.png", auto_pdf = TRUE)
```

## Graph-based clustering

Pros

  * fast and memory efficient (no distance matrix for all pairs of cells)
  * no assumptions on the shape of the clusters or the distribution of cells
  within each cluster
  * no need to specify a number of clusters to identify
  
Cons

  * loss of information beyond neighboring cells, which can affect community
  detection in regions with many cells.

## Modularity

Several methods to detect clusters (‘communities’) in networks rely on the
‘modularity’ metric.

Modularity measures how separated clusters are from each other.

Modularity is a ratio between the observed weights of the edges
within a cluster versus the expected weights if the edges were randomly
distributed between all nodes.

For the whole graph, the closer to 1 the better.

## Walktrap

The walktrap method relies on short random walks (a few steps) through the network.

These walks tend to be 'trapped' in highly-connected regions of the network.

Node similarity is measured based on these walks.

* Nodes are first each assigned their own community.
* Pairwise distances are computed and the two closest communities are grouped.
* These steps are repeated a given number of times to produce a dendrogram.
  * Hierarchical clustering is applied to the distance matrix.
* The best partition is that with the highest modularity.

<!-- https://link.springer.com/article/10.1007/s11227-019-03018-x -->

## Walktrap

Network example:

```{r, echo=FALSE, out.height='60%', out.width = '60%', fig.align="center"}
knitr::include_graphics("Images/clusGraphExample.png", auto_pdf = TRUE)
```

## Louvain {.columns-2 .smaller}

Hierarchical agglomerative method

Nodes are also first assigned their own community.

Two-step iterations:

* nodes are re-assigned one at a time to the community for which they increase modularity the most,
* a new, 'aggregate' network is built where nodes are the communities formed in the previous step.

This is repeated until modularity stops increasing.

([Blondel et al, Fast unfolding of communities in large networks](https://iopscience.iop.org/article/10.1088/1742-5468/2008/10/P10008/meta))

<p class="forceBreak"></p>

```{r, echo=FALSE, out.height='80%', out.width = '80%', fig.align="center"}
knitr::include_graphics("../Images/leiden_Fig1_noLegend.png", auto_pdf = TRUE)
```

([Traag et al, From Louvain to Leiden: guaranteeing well-connected communities](https://www.nature.com/articles/s41598-019-41695-z))

## Leiden

Issue with the Louvain method: some communities may be disconnected:

```{r, echo=FALSE, out.height='60%', out.width = '60%', fig.align="center"}
knitr::include_graphics("../Images/leiden_Fig2_HTML.png", auto_pdf = TRUE)
```

([Traag et al, From Louvain to Leiden: guaranteeing well-connected communities](https://www.nature.com/articles/s41598-019-41695-z))

## Leiden {.columns-2 .smaller}

The Leiden method improves on the Louvain method

by garanteeing that at each iteration clusters are connected and well-separated.

The method includes an extra step in the iterations:

* after nodes are moved (step 1),
* the resulting partition is refined (step2)
* and only then the new aggregate network made, and refined (step 3).

<p class="forceBreak"></p>

```{r, echo=FALSE, out.height='100%', out.width = '100%', fig.align="center"}
knitr::include_graphics("../Images/leiden_Fig3_noLegend.png", auto_pdf = TRUE)
```

([Traag et al, From Louvain to Leiden: guaranteeing well-connected communities](https://www.nature.com/articles/s41598-019-41695-z))

## Separatedness - silhouette width

Congruence of clusters may be assessed by computing the silhouette width for 
each cell.

For each cell in the cluster calculate the the average distance to all other
cells in the cluster and the average distance to all cells not in the cluster.
The cells silhouette width is the difference between these divided by the
maximum of the two values. 

Cells with a large silhouette are strongly related to cells in the cluster,
cells with a negative silhouette width are more closely related to other
clusters.

Good cluster separation is indicated by clusters whose cells have large
silhouette values.

## Separatedness - silhouette width

```{r, echo=FALSE, out.width = '100%', fig.align="center"}
knitr::include_graphics("Images/Silhouette.png")
```

## Cluster-wise modularity to assess clusters quality 

Clusters that are well separated mostly comprise intra-cluster edges and harbour
a high modularity score on the diagonal and low scores off that diagonal.

Two poorly separated clusters will share edges and the pair will have a high score.

```{r, echo=FALSE, out.width = '90%', fig.align="center"}
knitr::include_graphics("Images/ClusterwiseModularity.png")
```
