<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.529">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Introduction to single-cell RNA-seq analysis</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script><script src="scanpy_part1_files/libs/clipboard/clipboard.min.js"></script>
<script src="scanpy_part1_files/libs/quarto-html/quarto.js"></script>
<script src="scanpy_part1_files/libs/quarto-html/popper.min.js"></script>
<script src="scanpy_part1_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="scanpy_part1_files/libs/quarto-html/anchor.min.js"></script>
<link href="scanpy_part1_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="scanpy_part1_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="scanpy_part1_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="scanpy_part1_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="scanpy_part1_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" integrity="sha512-c3Nl8+7g4LMSTdrm621y7kf9v3SDPnhxLNhcjFJbKECVnmZHTdo+IRO05sNLTH/D3vA6u1X32ehoLC7WFVdheg==" crossorigin="anonymous"></script>

<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>


</head>

<body>

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article toc-left">
<div id="quarto-sidebar-toc-left" class="sidebar toc-left">
  <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#setting-up" id="toc-setting-up" class="nav-link active" data-scroll-target="#setting-up"><span class="header-section-number">1</span> Setting up</a></li>
  <li><a href="#reading-data" id="toc-reading-data" class="nav-link" data-scroll-target="#reading-data"><span class="header-section-number">2</span> Reading data</a></li>
  <li><a href="#gene-annotation" id="toc-gene-annotation" class="nav-link" data-scroll-target="#gene-annotation"><span class="header-section-number">3</span> Gene annotation</a></li>
  <li><a href="#filtering" id="toc-filtering" class="nav-link" data-scroll-target="#filtering"><span class="header-section-number">4</span> Filtering</a>
  <ul class="collapse">
  <li><a href="#filtering-barcodes" id="toc-filtering-barcodes" class="nav-link" data-scroll-target="#filtering-barcodes"><span class="header-section-number">4.1</span> Filtering barcodes</a></li>
  <li><a href="#filtering-genes" id="toc-filtering-genes" class="nav-link" data-scroll-target="#filtering-genes"><span class="header-section-number">4.2</span> Filtering genes</a></li>
  <li><a href="#ambient-rna-and-doublets" id="toc-ambient-rna-and-doublets" class="nav-link" data-scroll-target="#ambient-rna-and-doublets"><span class="header-section-number">4.3</span> Ambient RNA and doublets</a></li>
  </ul></li>
  <li><a href="#normalisation" id="toc-normalisation" class="nav-link" data-scroll-target="#normalisation"><span class="header-section-number">5</span> Normalisation</a></li>
  <li><a href="#preprocessing-workflow" id="toc-preprocessing-workflow" class="nav-link" data-scroll-target="#preprocessing-workflow"><span class="header-section-number">6</span> Preprocessing workflow</a></li>
  </ul>
</nav>
</div>
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
</div>
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Introduction to single-cell RNA-seq analysis</h1>
<p class="subtitle lead">Python/scanpy workflow</p>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<section id="setting-up" class="level2" data-number="1">
<h2 data-number="1" class="anchored" data-anchor-id="setting-up"><span class="header-section-number">1</span> Setting up</h2>
<div id="b0e97a16" class="cell" data-execution_count="2">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> scipy.stats <span class="im">import</span> median_abs_deviation</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> scipy.sparse <span class="im">import</span> csr_matrix</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> scanpy <span class="im">as</span> sc</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> anndata <span class="im">as</span> ad</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pybiomart <span class="im">as</span> bm</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> seaborn <span class="im">as</span> sns</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="co"># set plotting theme</span></span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a>sns.set_theme()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We are working with data from <a href="https://doi.org/10.1038/s41598-020-64929-x">Caron et al.&nbsp;2020</a>, investigating the transcriptional profile of different types of childhood acute lymphoblastic leukemia (ALL). There are different subtypes of ALL, with the two main groups being B-cell lymphoblastic leukemia (affecting immature B-cell lymphocytes) and T-cell lymphoblastic leukemia (affecting immature T-cell lymphocytes). Within each of these broad types, there are further classifications based on commonly observed genetic changes such as large chromosomal rearragements.</p>
<p>In this dataset (<a href="https://www.ncbi.nlm.nih.gov/bioproject/?term=PRJNA548203">SRA Accession PRJNA548203</a>), we have 4 groups of samples:</p>
<ul>
<li><code>ETV6-RUNX1</code> (4 replicates): samples from a type of B-cell precursor ALL characterised by a translocation between chromosomes 12 and 21 known as the <em>t(12;21) ETV6/RUNX1</em> rearrangement.</li>
<li><code>HHD</code> (2 replicates): also a type of B-cell precursor ALL, where extra chromosomes are gained and known as the “high hyper diploid” type.</li>
<li><code>PRE-T</code> (2 replicates): a type of T-cell precursor ALL.</li>
<li><code>PBMMC</code> (3 replicates): control samples from healthy pediatric bone marrow mononuclear cells.</li>
</ul>
<p>We will start by showing how to read and preprocessing the data for one of these samples only. We will then show how to apply the preprocessing to all samples, after which we integrate them together taking into account batch effects inherent to these type of data.</p>
</section>
<section id="reading-data" class="level2" data-number="2">
<h2 data-number="2" class="anchored" data-anchor-id="reading-data"><span class="header-section-number">2</span> Reading data</h2>
<p>We start by reading the data using one of the read functions from <code>scanpy</code> (<a href="https://scanpy.readthedocs.io/en/stable/api/reading.html">documented here</a>). In our case, we have our data as matrices from CellRanger, so we use the function <code>sc.read_10x_mtx()</code>, which takes a directory as its input. We read one of the replicates of the ETV6/RUNX1 subtype:</p>
<div id="e56cb31a" class="cell" data-execution_count="4">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>ETV6_RUNX1_1 <span class="op">=</span> sc.read_10x_mtx(<span class="st">"Data/CellRanger_Outputs/SRR9264343/outs/filtered_feature_bc_matrix/"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This returns an object of class <code>AnnData</code>:</p>
<div id="2525cbb6" class="cell" data-execution_count="5">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>ETV6_RUNX1_1</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="4">
<pre><code>AnnData object with n_obs × n_vars = 3153 × 37764
    var: 'gene_ids', 'feature_types'</code></pre>
</div>
</div>
<p>This object stores several pieces of information, including:</p>
<ul>
<li>A matrix of raw counts, with samples as rows and genes as columns. This can be retrived with the <code>.X</code> accessor.</li>
<li>Metadata about the barcodes (cells), which can be retrieved using the <code>.obs</code> accessor.</li>
<li>Metadata about the genes, which can be retrived using the <code>.var</code> accessor.</li>
</ul>
<p>In the output above, we get indication of the number of barcodes (<code>n_obs</code>) and genes (<code>n_vars</code>).</p>
<div class="img-fluid quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="https://falexwolf.de/img/scanpy/anndata.svg" class="img-fluid figure-img"></p>
<figcaption>Diagram of the AnnData object type. <a href="https://scanpy.readthedocs.io/en/stable/usage-principles.html">Source</a></figcaption>
</figure>
</div>
</section>
<section id="gene-annotation" class="level2" data-number="3">
<h2 data-number="3" class="anchored" data-anchor-id="gene-annotation"><span class="header-section-number">3</span> Gene annotation</h2>
<div id="74d7a172" class="cell" data-execution_count="6">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>ETV6_RUNX1_1.var</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="5">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">gene_ids</th>
<th data-quarto-table-cell-role="th">feature_types</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">MIR1302-2HG</td>
<td>ENSG00000243485</td>
<td>Gene Expression</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">FAM138A</td>
<td>ENSG00000237613</td>
<td>Gene Expression</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">OR4F5</td>
<td>ENSG00000186092</td>
<td>Gene Expression</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">ENSG00000238009</td>
<td>ENSG00000238009</td>
<td>Gene Expression</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">ENSG00000239945</td>
<td>ENSG00000239945</td>
<td>Gene Expression</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">...</td>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">ENSG00000277836</td>
<td>ENSG00000277836</td>
<td>Gene Expression</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">ENSG00000278633</td>
<td>ENSG00000278633</td>
<td>Gene Expression</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">ENSG00000276017</td>
<td>ENSG00000276017</td>
<td>Gene Expression</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">ENSG00000278817</td>
<td>ENSG00000278817</td>
<td>Gene Expression</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">ENSG00000277196</td>
<td>ENSG00000277196</td>
<td>Gene Expression</td>
</tr>
</tbody>
</table>

<p>37764 rows × 2 columns</p>
</div>
</div>
</div>
<p>We can see that our gene ids use ENSEMBL identifiers. For each of these identifiers, we would like to know information about common gene names and also the chromosome, so we can identify mitochondrial genes.</p>
<p>We will use the <code>pybiomart</code> package, which accesses the Biomart database containing information about genes for many species. You can learn more about the package usage from <a href="https://jrderuiter.github.io/pybiomart/usage.html">its documentation</a>.</p>
<div id="12b690f0" class="cell" data-execution_count="8">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># connect to the Human genes database (GRCh38.p14)</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>h38_mart <span class="op">=</span> bm.Dataset(name<span class="op">=</span><span class="st">"hsapiens_gene_ensembl"</span>,</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>                      host<span class="op">=</span><span class="st">"http://www.ensembl.org"</span>)</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a><span class="co"># retrieve gene information</span></span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>h38_genes <span class="op">=</span> h38_mart.query(attributes<span class="op">=</span>[<span class="st">"ensembl_gene_id"</span>, <span class="st">"external_gene_name"</span>, <span class="st">"chromosome_name"</span>])</span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a><span class="co"># rename the columns</span></span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a>h38_genes <span class="op">=</span> h38_genes.rename(columns<span class="op">=</span>{<span class="st">"Gene stable ID"</span>: <span class="st">"gene_ids"</span>, <span class="st">"Gene name"</span>: <span class="st">"gene_name"</span>, <span class="st">"Chromosome/scaffold name"</span>: <span class="st">"chrom"</span>})</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This returns a Pandas DataFrame object:</p>
<div id="92b84e90" class="cell" data-execution_count="9">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>h38_genes</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="7">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">gene_ids</th>
<th data-quarto-table-cell-role="th">gene_name</th>
<th data-quarto-table-cell-role="th">chrom</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>ENSG00000210049</td>
<td>MT-TF</td>
<td>MT</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>ENSG00000211459</td>
<td>MT-RNR1</td>
<td>MT</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>ENSG00000210077</td>
<td>MT-TV</td>
<td>MT</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>ENSG00000210082</td>
<td>MT-RNR2</td>
<td>MT</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>ENSG00000209082</td>
<td>MT-TL1</td>
<td>MT</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">...</td>
<td>...</td>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">70706</td>
<td>ENSG00000288629</td>
<td>NaN</td>
<td>1</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">70707</td>
<td>ENSG00000288678</td>
<td>NaN</td>
<td>1</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">70708</td>
<td>ENSG00000290825</td>
<td>DDX11L2</td>
<td>1</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">70709</td>
<td>ENSG00000227232</td>
<td>WASH7P</td>
<td>1</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">70710</td>
<td>ENSG00000290826</td>
<td>NaN</td>
<td>1</td>
</tr>
</tbody>
</table>

<p>70711 rows × 3 columns</p>
</div>
</div>
</div>
<p>We can now merge this DataFrame with the DataFrame from our AnnData metadata:</p>
<div id="52ab2a66" class="cell" data-execution_count="10">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>gene_annot <span class="op">=</span> (</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>  ETV6_RUNX1_1.var</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>  .merge(h38_genes, how<span class="op">=</span><span class="st">"left"</span>, on<span class="op">=</span><span class="st">"gene_ids"</span>)</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>  .set_axis(ETV6_RUNX1_1.var.index)</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Finally, we re-annotate our AnnData, being careful to ensure the order of our genes is the same as in the original AnnData object:</p>
<div id="9d34607c" class="cell" data-execution_count="11">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>ETV6_RUNX1_1.var <span class="op">=</span> gene_annot.loc[ETV6_RUNX1_1.var_names]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We will keep only genes in the autosomes, X, Y and MT chromosome (i.e.&nbsp;remove genes in unassembled scaffolds):</p>
<div id="70d8f202" class="cell" data-execution_count="12">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>vars_to_keep <span class="op">=</span> (</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>  ETV6_RUNX1_1.var[<span class="st">"chrom"</span>]</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>  .isin([<span class="bu">str</span>(i) <span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">1</span>, <span class="dv">23</span>)] <span class="op">+</span> [<span class="st">"X"</span>, <span class="st">"Y"</span>, <span class="st">"MT"</span>])</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>ETV6_RUNX1_1 <span class="op">=</span> ETV6_RUNX1_1[:, vars_to_keep].copy()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Note the use of the <code>.copy()</code> method. This ensures that we make a new copy of the object (which we replace back into <code>ETV6_RUNX1_1</code>), rather than a “View” of the original object.</p>
<p>Finally, we create a variable in our gene metadata to indicate whether a gene is mitochondrial or not. We will use this later on when we do quality control.</p>
<div id="ce9c41ac" class="cell" data-execution_count="13">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>ETV6_RUNX1_1.var[<span class="st">"mt"</span>] <span class="op">=</span> ETV6_RUNX1_1.var[<span class="st">"chrom"</span>] <span class="op">==</span> <span class="st">"MT"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="filtering" class="level2" data-number="4">
<h2 data-number="4" class="anchored" data-anchor-id="filtering"><span class="header-section-number">4</span> Filtering</h2>
<p>We start by doing some exploratory analysis of our raw count data, namely in terms of:</p>
<ul>
<li>number of total counts per barcode</li>
<li>number of detected genes per barcode</li>
<li>fraction of counts in mitochondrial genes</li>
</ul>
<div id="9e84fa3c" class="cell" data-execution_count="14">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>sc.pp.calculate_qc_metrics(</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>    ETV6_RUNX1_1, </span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>    qc_vars<span class="op">=</span>[<span class="st">"mt"</span>], </span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>    inplace<span class="op">=</span><span class="va">True</span>, </span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>    percent_top<span class="op">=</span>[<span class="dv">20</span>], </span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>    log1p<span class="op">=</span><span class="va">True</span></span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a>ETV6_RUNX1_1</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="12">
<pre><code>AnnData object with n_obs × n_vars = 3153 × 37663
    obs: 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'pct_counts_in_top_20_genes', 'total_counts_mt', 'log1p_total_counts_mt', 'pct_counts_mt'
    var: 'gene_ids', 'feature_types', 'gene_name', 'chrom', 'mt', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts'</code></pre>
</div>
</div>
<p>This function added several metrics for each barcode, i.e.&nbsp;our observations:</p>
<div id="67ef93f0" class="cell" data-execution_count="15">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>ETV6_RUNX1_1.obs</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="13">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">n_genes_by_counts</th>
<th data-quarto-table-cell-role="th">log1p_n_genes_by_counts</th>
<th data-quarto-table-cell-role="th">total_counts</th>
<th data-quarto-table-cell-role="th">log1p_total_counts</th>
<th data-quarto-table-cell-role="th">pct_counts_in_top_20_genes</th>
<th data-quarto-table-cell-role="th">total_counts_mt</th>
<th data-quarto-table-cell-role="th">log1p_total_counts_mt</th>
<th data-quarto-table-cell-role="th">pct_counts_mt</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">AAACCTGAGACTTTCG-1</td>
<td>2919</td>
<td>7.979339</td>
<td>8314.0</td>
<td>9.025816</td>
<td>14.662016</td>
<td>293.0</td>
<td>5.683580</td>
<td>3.524176</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">AAACCTGGTCTTCAAG-1</td>
<td>4317</td>
<td>8.370548</td>
<td>14914.0</td>
<td>9.610123</td>
<td>14.851817</td>
<td>570.0</td>
<td>6.347389</td>
<td>3.821912</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">AAACCTGGTGCAACTT-1</td>
<td>931</td>
<td>6.837333</td>
<td>1567.0</td>
<td>7.357556</td>
<td>31.014678</td>
<td>431.0</td>
<td>6.068426</td>
<td>27.504787</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">AAACCTGGTGTTGAGG-1</td>
<td>3626</td>
<td>8.196161</td>
<td>10429.0</td>
<td>9.252441</td>
<td>15.591140</td>
<td>427.0</td>
<td>6.059123</td>
<td>4.094352</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">AAACCTGTCCCAAGTA-1</td>
<td>3317</td>
<td>8.107117</td>
<td>10407.0</td>
<td>9.250330</td>
<td>17.199962</td>
<td>528.0</td>
<td>6.270988</td>
<td>5.073508</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">TTTGGTTTCTTTAGGG-1</td>
<td>2485</td>
<td>7.818430</td>
<td>5029.0</td>
<td>8.523175</td>
<td>14.495924</td>
<td>238.0</td>
<td>5.476463</td>
<td>4.732551</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">TTTGTCAAGAAACGAG-1</td>
<td>3430</td>
<td>8.140607</td>
<td>9864.0</td>
<td>9.196749</td>
<td>14.355231</td>
<td>549.0</td>
<td>6.309918</td>
<td>5.565693</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">TTTGTCAAGGACGAAA-1</td>
<td>3713</td>
<td>8.219865</td>
<td>11776.0</td>
<td>9.373903</td>
<td>14.393682</td>
<td>498.0</td>
<td>6.212606</td>
<td>4.228940</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">TTTGTCACAGGCTCAC-1</td>
<td>1257</td>
<td>7.137278</td>
<td>2098.0</td>
<td>7.649216</td>
<td>25.786463</td>
<td>225.0</td>
<td>5.420535</td>
<td>10.724500</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">TTTGTCAGTTCGGCAC-1</td>
<td>4057</td>
<td>8.308446</td>
<td>13111.0</td>
<td>9.481283</td>
<td>15.323011</td>
<td>486.0</td>
<td>6.188264</td>
<td>3.706811</td>
</tr>
</tbody>
</table>

<p>3153 rows × 8 columns</p>
</div>
</div>
</div>
<p>And also to our genes, i.e.&nbsp;variables:</p>
<div id="79047279" class="cell" data-execution_count="16">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>ETV6_RUNX1_1.var</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="14">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">gene_ids</th>
<th data-quarto-table-cell-role="th">feature_types</th>
<th data-quarto-table-cell-role="th">gene_name</th>
<th data-quarto-table-cell-role="th">chrom</th>
<th data-quarto-table-cell-role="th">mt</th>
<th data-quarto-table-cell-role="th">n_cells_by_counts</th>
<th data-quarto-table-cell-role="th">mean_counts</th>
<th data-quarto-table-cell-role="th">log1p_mean_counts</th>
<th data-quarto-table-cell-role="th">pct_dropout_by_counts</th>
<th data-quarto-table-cell-role="th">total_counts</th>
<th data-quarto-table-cell-role="th">log1p_total_counts</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">MIR1302-2HG</td>
<td>ENSG00000243485</td>
<td>Gene Expression</td>
<td>MIR1302-2HG</td>
<td>1</td>
<td>False</td>
<td>0</td>
<td>0.000000</td>
<td>0.000000</td>
<td>100.000000</td>
<td>0.0</td>
<td>0.000000</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">FAM138A</td>
<td>ENSG00000237613</td>
<td>Gene Expression</td>
<td>FAM138A</td>
<td>1</td>
<td>False</td>
<td>0</td>
<td>0.000000</td>
<td>0.000000</td>
<td>100.000000</td>
<td>0.0</td>
<td>0.000000</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">OR4F5</td>
<td>ENSG00000186092</td>
<td>Gene Expression</td>
<td>OR4F5</td>
<td>1</td>
<td>False</td>
<td>0</td>
<td>0.000000</td>
<td>0.000000</td>
<td>100.000000</td>
<td>0.0</td>
<td>0.000000</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">ENSG00000238009</td>
<td>ENSG00000238009</td>
<td>Gene Expression</td>
<td>NaN</td>
<td>1</td>
<td>False</td>
<td>4</td>
<td>0.001269</td>
<td>0.001268</td>
<td>99.873137</td>
<td>4.0</td>
<td>1.609438</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">ENSG00000239945</td>
<td>ENSG00000239945</td>
<td>Gene Expression</td>
<td>NaN</td>
<td>1</td>
<td>False</td>
<td>0</td>
<td>0.000000</td>
<td>0.000000</td>
<td>100.000000</td>
<td>0.0</td>
<td>0.000000</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">MT-ND4L</td>
<td>ENSG00000212907</td>
<td>Gene Expression</td>
<td>MT-ND4L</td>
<td>MT</td>
<td>True</td>
<td>2456</td>
<td>2.021250</td>
<td>1.105670</td>
<td>22.105931</td>
<td>6373.0</td>
<td>8.759982</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">MT-ND4</td>
<td>ENSG00000198886</td>
<td>Gene Expression</td>
<td>MT-ND4</td>
<td>MT</td>
<td>True</td>
<td>3126</td>
<td>27.849033</td>
<td>3.362077</td>
<td>0.856327</td>
<td>87808.0</td>
<td>11.382919</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">MT-ND5</td>
<td>ENSG00000198786</td>
<td>Gene Expression</td>
<td>MT-ND5</td>
<td>MT</td>
<td>True</td>
<td>2884</td>
<td>4.335236</td>
<td>1.674333</td>
<td>8.531557</td>
<td>13669.0</td>
<td>9.522959</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">MT-ND6</td>
<td>ENSG00000198695</td>
<td>Gene Expression</td>
<td>MT-ND6</td>
<td>MT</td>
<td>True</td>
<td>1981</td>
<td>1.843007</td>
<td>1.044862</td>
<td>37.170948</td>
<td>5811.0</td>
<td>8.667680</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">MT-CYB</td>
<td>ENSG00000198727</td>
<td>Gene Expression</td>
<td>MT-CYB</td>
<td>MT</td>
<td>True</td>
<td>3092</td>
<td>15.561053</td>
<td>2.807054</td>
<td>1.934665</td>
<td>49064.0</td>
<td>10.800901</td>
</tr>
</tbody>
</table>

<p>37663 rows × 11 columns</p>
</div>
</div>
</div>
<section id="filtering-barcodes" class="level3" data-number="4.1">
<h3 data-number="4.1" class="anchored" data-anchor-id="filtering-barcodes"><span class="header-section-number">4.1</span> Filtering barcodes</h3>
<p>Since <code>ETV6_RUNX1_1.obs</code> is a regular DataFrame, we can use standard plotting libraries to visualise these statistics. For example, using the popular Seaborn library:</p>
<div id="9ca0bcd1" class="cell" data-execution_count="17">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>sns.displot(ETV6_RUNX1_1.obs, x<span class="op">=</span><span class="st">"total_counts"</span>, bins<span class="op">=</span><span class="dv">100</span>)</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>sns.displot(ETV6_RUNX1_1.obs, x<span class="op">=</span><span class="st">"pct_counts_mt"</span>, bins<span class="op">=</span><span class="dv">100</span>)</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>sns.scatterplot(ETV6_RUNX1_1.obs, x<span class="op">=</span><span class="st">"total_counts"</span>, y<span class="op">=</span><span class="st">"n_genes_by_counts"</span>, hue<span class="op">=</span><span class="st">"pct_counts_mt"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="scanpy_part1_files/figure-html/cell-17-output-1.png" width="465" height="465" class="figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="scanpy_part1_files/figure-html/cell-17-output-2.png" width="490" height="462" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p>Alternatively, we can use <code>scanpy</code>’s own plotting functions (histogram is not available, but we can do violin plots instead):</p>
<div id="52cb10b2" class="cell" data-execution_count="18">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>sc.pl.violin(ETV6_RUNX1_1, <span class="st">"total_counts"</span>)</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>sc.pl.violin(ETV6_RUNX1_1, <span class="st">"pct_counts_mt"</span>)</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>sc.pl.scatter(ETV6_RUNX1_1, <span class="st">"total_counts"</span>, <span class="st">"n_genes_by_counts"</span>, color<span class="op">=</span><span class="st">"pct_counts_mt"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="scanpy_part1_files/figure-html/cell-18-output-1.png" width="683" height="397" class="figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="scanpy_part1_files/figure-html/cell-18-output-2.png" width="655" height="397" class="figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="scanpy_part1_files/figure-html/cell-18-output-3.png" width="743" height="438" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p>We can even do several violin plots at once:</p>
<div id="6f51b056" class="cell" data-execution_count="19">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>sc.pl.violin(</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>  ETV6_RUNX1_1,</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>  [<span class="st">"n_genes_by_counts"</span>, <span class="st">"total_counts"</span>, <span class="st">"pct_counts_mt"</span>],</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>  multi_panel<span class="op">=</span><span class="va">True</span></span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="scanpy_part1_files/figure-html/cell-19-output-1.png" width="1446" height="465" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p>We can filter our object based on hard thresholds set manually. Alternatively, we can define a function that removes outliers based on the observed distribution:</p>
<div id="71e3b736" class="cell" data-execution_count="20">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> is_outlier(adata, metric: <span class="bu">str</span>, nmads: <span class="bu">int</span>):</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>  M <span class="op">=</span> adata.obs[metric]</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>  outlier <span class="op">=</span> (M <span class="op">&lt;</span> np.median(M) <span class="op">-</span> nmads <span class="op">*</span> median_abs_deviation(M)) <span class="op">|</span> (</span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a>      np.median(M) <span class="op">+</span> nmads <span class="op">*</span> median_abs_deviation(M) <span class="op">&lt;</span> M</span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a>  <span class="cf">return</span> outlier</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The function returns <code>True</code> or <code>False</code> depending on whether the value exceeds the specified value of median absolute deviation. For example:</p>
<div id="f70235d6" class="cell" data-execution_count="21">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="co"># create barcode metadata column indicating counts outliers</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>ETV6_RUNX1_1.obs[<span class="st">"counts_outlier"</span>] <span class="op">=</span> is_outlier(ETV6_RUNX1_1, <span class="st">"log1p_total_counts"</span>, <span class="dv">5</span>)</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a><span class="co"># visualise</span></span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>sns.displot(ETV6_RUNX1_1.obs, x<span class="op">=</span><span class="st">"log1p_total_counts"</span>, hue<span class="op">=</span><span class="st">"counts_outlier"</span>, multiple<span class="op">=</span><span class="st">"stack"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="scanpy_part1_files/figure-html/cell-21-output-1.png" width="600" height="465" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p>Note that we used the log-transformed counts, as its distribution is less skewed and therefore more suitable for the MAD-based filtering we are doing.</p>
<p>We can repeat this for number of detected genes (also log-transformed) and the percentage of counts in the two 20 genes:</p>
<div id="3f1d2716" class="cell" data-execution_count="22">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>ETV6_RUNX1_1.obs[<span class="st">"genes_outlier"</span>] <span class="op">=</span> is_outlier(ETV6_RUNX1_1, <span class="st">"log1p_n_genes_by_counts"</span>, <span class="dv">5</span>)</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>sns.displot(ETV6_RUNX1_1.obs, x<span class="op">=</span><span class="st">"log1p_n_genes_by_counts"</span>, hue<span class="op">=</span><span class="st">"genes_outlier"</span>, multiple<span class="op">=</span><span class="st">"stack"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="scanpy_part1_files/figure-html/cell-22-output-1.png" width="595" height="465" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="dbd6d8f0" class="cell" data-execution_count="23">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>ETV6_RUNX1_1.obs[<span class="st">"topgenes_outlier"</span>] <span class="op">=</span> is_outlier(ETV6_RUNX1_1, <span class="st">"pct_counts_in_top_20_genes"</span>, <span class="dv">5</span>)</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>sns.displot(ETV6_RUNX1_1.obs, x<span class="op">=</span><span class="st">"pct_counts_in_top_20_genes"</span>, hue<span class="op">=</span><span class="st">"topgenes_outlier"</span>, multiple<span class="op">=</span><span class="st">"stack"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="scanpy_part1_files/figure-html/cell-23-output-1.png" width="621" height="465" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p>We also check for outliers with regards to percentage of mitochondrial counts, where we use more strict filters:</p>
<div id="ada1f36f" class="cell" data-execution_count="24">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>ETV6_RUNX1_1.obs[<span class="st">"mito_outlier"</span>] <span class="op">=</span> is_outlier(ETV6_RUNX1_1, <span class="st">"pct_counts_mt"</span>, <span class="dv">3</span>) <span class="op">|</span> (ETV6_RUNX1_1.obs[<span class="st">"pct_counts_mt"</span>] <span class="op">&gt;</span> <span class="dv">8</span>)</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>sns.displot(ETV6_RUNX1_1.obs, x<span class="op">=</span><span class="st">"pct_counts_mt"</span>, hue<span class="op">=</span><span class="st">"mito_outlier"</span>, multiple<span class="op">=</span><span class="st">"stack"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="scanpy_part1_files/figure-html/cell-24-output-1.png" width="583" height="465" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p>Finally, we create a variable which is the union of these conditions, i.e.&nbsp;if the barcode is determined an outlier of <em>any</em> of our filters, then we consider it to be an outlier:</p>
<div id="a7a52d88" class="cell" data-execution_count="25">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>ETV6_RUNX1_1.obs[<span class="st">"outlier"</span>] <span class="op">=</span> ETV6_RUNX1_1.obs[<span class="st">"genes_outlier"</span>] <span class="op">|</span> ETV6_RUNX1_1.obs[<span class="st">"genes_outlier"</span>] <span class="op">|</span> ETV6_RUNX1_1.obs[<span class="st">"topgenes_outlier"</span>] <span class="op">|</span> ETV6_RUNX1_1.obs[<span class="st">"mito_outlier"</span>]</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>ETV6_RUNX1_1.obs[<span class="st">"outlier"</span>].value_counts()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="23">
<pre><code>outlier
False    2674
True      479
Name: count, dtype: int64</code></pre>
</div>
</div>
<p>We can visualise our scatterplot of counts vs detected genes to see which barcodes will be removed:</p>
<div id="f7843215" class="cell" data-execution_count="26">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>sns.scatterplot(ETV6_RUNX1_1.obs, </span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>                x <span class="op">=</span> <span class="st">"total_counts"</span>, </span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>                y <span class="op">=</span> <span class="st">"n_genes_by_counts"</span>,</span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a>                hue <span class="op">=</span> <span class="st">"outlier"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="scanpy_part1_files/figure-html/cell-26-output-1.png" width="624" height="437" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p>Finally, we remove these barcodes from downstream analysis (note the <code>~</code> operator <em>negates</em> the True/False variable, as we want to keep the barcodes where the “outlier” column is False):</p>
<div id="44783c05" class="cell" data-execution_count="27">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>ETV6_RUNX1_1 <span class="op">=</span> ETV6_RUNX1_1[<span class="op">~</span>ETV6_RUNX1_1.obs[<span class="st">"outlier"</span>], :].copy()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The barcodes we are left with after this filtering we will now consider to be cells.</p>
</section>
<section id="filtering-genes" class="level3" data-number="4.2">
<h3 data-number="4.2" class="anchored" data-anchor-id="filtering-genes"><span class="header-section-number">4.2</span> Filtering genes</h3>
<p>In the same way that we explored several metrics for barcodes, we can also explore them for genes. However, as we will see, downstream analysis can focus on variable genes and will mostly ignore genes for which there is very little data. Therefore, we don’t perform as strict filtering on genes as we do on barcodes.</p>
<p>Still, it is useful to remove undetected genes, i.e.&nbsp;those with zero total counts:</p>
<div id="7fc05584" class="cell" data-execution_count="28">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="co"># count of genes with zero counts</span></span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>ETV6_RUNX1_1.var[<span class="st">"total_counts"</span>].eq(<span class="dv">0</span>).value_counts()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="26">
<pre><code>total_counts
False    26436
True     11227
Name: count, dtype: int64</code></pre>
</div>
</div>
<p>We can use the <code>sc.pp.filter_genes()</code> function to do this</p>
<div id="1f27e4c3" class="cell" data-execution_count="29">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>sc.pp.filter_genes(ETV6_RUNX1_1, min_counts<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a><span class="co"># check number  of genes left</span></span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a>ETV6_RUNX1_1.n_vars</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="27">
<pre><code>37663</code></pre>
</div>
</div>
</section>
<section id="ambient-rna-and-doublets" class="level3" data-number="4.3">
<h3 data-number="4.3" class="anchored" data-anchor-id="ambient-rna-and-doublets"><span class="header-section-number">4.3</span> Ambient RNA and doublets</h3>
<p>TODO</p>
</section>
</section>
<section id="normalisation" class="level2" data-number="5">
<h2 data-number="5" class="anchored" data-anchor-id="normalisation"><span class="header-section-number">5</span> Normalisation</h2>
<p>There are several normalisation methods available for single-cell RNA-seq data. Two of those methods are:</p>
<ul>
<li><strong>Pearson residuals:</strong> this method models the count data using a negative binomial regression model to regress out the effects of total count differences between cells and uses the residuals from the model as the normalised count values.</li>
<li><strong>Shifted logarithm:</strong> scales the counts by a cell-specific size factor (based on the total counts in that cell) followed by taking its logarithm. Despite being a relatively simple method, it has been shown to perform well in downstream analysis such as dimensionality reduction and clustering.</li>
</ul>
<p>We will use the <code>layers</code> component of the AnnData object, which can be used to store different versions of our count matrix. This is a good way to keep several versions of our data in place, especially as we explore different methods of normalisation.</p>
<div id="6404f4ea" class="cell" data-execution_count="30">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="co"># keep a copy of the raw counts in the object as a backup</span></span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>ETV6_RUNX1_1.layers[<span class="st">"counts"</span>] <span class="op">=</span> ETV6_RUNX1_1.X.copy()</span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a><span class="co"># create a new layer for Pearson residuals</span></span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a>ETV6_RUNX1_1.layers[<span class="st">"pearson"</span>] <span class="op">=</span> ETV6_RUNX1_1.X.copy()</span>
<span id="cb32-6"><a href="#cb32-6" aria-hidden="true" tabindex="-1"></a>sc.experimental.pp.normalize_pearson_residuals(ETV6_RUNX1_1, layer<span class="op">=</span><span class="st">"pearson"</span>)</span>
<span id="cb32-7"><a href="#cb32-7" aria-hidden="true" tabindex="-1"></a><span class="co"># convert to sparse matrix (efficient and saves memory)</span></span>
<span id="cb32-8"><a href="#cb32-8" aria-hidden="true" tabindex="-1"></a>ETV6_RUNX1_1.layers[<span class="st">"pearson"</span>] <span class="op">=</span> csr_matrix(ETV6_RUNX1_1.layers[<span class="st">"pearson"</span>])</span>
<span id="cb32-9"><a href="#cb32-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-10"><a href="#cb32-10" aria-hidden="true" tabindex="-1"></a><span class="co"># create new layer for log-normalised counts</span></span>
<span id="cb32-11"><a href="#cb32-11" aria-hidden="true" tabindex="-1"></a>ETV6_RUNX1_1.layers[<span class="st">"logcounts"</span>] <span class="op">=</span> ETV6_RUNX1_1.X.copy()</span>
<span id="cb32-12"><a href="#cb32-12" aria-hidden="true" tabindex="-1"></a>sc.pp.normalize_total(ETV6_RUNX1_1, layer<span class="op">=</span><span class="st">"logcounts"</span>, target_sum<span class="op">=</span><span class="va">None</span>)</span>
<span id="cb32-13"><a href="#cb32-13" aria-hidden="true" tabindex="-1"></a>sc.pp.log1p(ETV6_RUNX1_1, layer<span class="op">=</span><span class="st">"logcounts"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>/home/hugot/mambaforge/envs/scverse/lib/python3.11/site-packages/scanpy/experimental/pp/_normalization.py:68: RuntimeWarning: invalid value encountered in divide
  residuals = diff / np.sqrt(mu + mu**2 / theta)</code></pre>
</div>
</div>
<p>We can visualise the distributions and correlations of the normalised data as a pair plot:</p>
<div id="6cce000d" class="cell" data-execution_count="31">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="co"># note: np.nansum() only works in regular matrix, so we coerce from the sparse matrix format</span></span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>sns.pairplot(</span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a>  pd.DataFrame({<span class="st">"Raw counts"</span>: np.nansum(ETV6_RUNX1_1.layers[<span class="st">"counts"</span>].toarray(), <span class="dv">1</span>),</span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a>              <span class="st">"Log-normalised counts"</span>: np.nansum(ETV6_RUNX1_1.layers[<span class="st">"logcounts"</span>].toarray(), <span class="dv">1</span>),</span>
<span id="cb34-5"><a href="#cb34-5" aria-hidden="true" tabindex="-1"></a>              <span class="st">"Pearson residuals"</span>: np.nansum(ETV6_RUNX1_1.layers[<span class="st">"pearson"</span>].toarray(), <span class="dv">1</span>)})</span>
<span id="cb34-6"><a href="#cb34-6" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="scanpy_part1_files/figure-html/cell-31-output-1.png" width="709" height="710" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p>As we can see, while the log-normalised data is fairly well correlated with the original counts, the Pearson residuals are not. This is because the Pearson residuals represent the output of a model taking into account the total counts of the cell, thus the values represent whether the expression is above or below the expected value of expression for that gene.</p>
<p>There is no “right method” when it comes to normalisation, and often several methods should be explored, alongside a knowledge of the biology. The Pearson residuals normalisation is argued to capture more of the relevant biological signal (<a href="https://doi.org/10.1186/s13059-021-02451-7">Lause et al.&nbsp;2021</a>) and the <a href="https://scanpy.readthedocs.io/en/stable/tutorials/experimental/pearson_residuals.html">Scanpy documentation</a> gives details about using this normalisation for downstream analysis.</p>
<p>However, as this method is computationally more demanding, we will proceed with the log normalisation.</p>
</section>
<section id="preprocessing-workflow" class="level2" data-number="6">
<h2 data-number="6" class="anchored" data-anchor-id="preprocessing-workflow"><span class="header-section-number">6</span> Preprocessing workflow</h2>
<p>Now that we have explored these things with a single sample, we can bring it all together to process all our samples. We will define a function that performs the preprocessing steps covered:</p>
<ul>
<li>Read the CellRanger matrix</li>
<li>Add gene annotation</li>
<li>Filter barcodes based on different metrics</li>
<li>Normalise the counts (using the simpler log-normalisation)</li>
</ul>
<div id="efffa475" class="cell" data-execution_count="32">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> sc_preprocess(<span class="bu">file</span>: <span class="bu">str</span>, gene_annot: pd.DataFrame):</span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a>  <span class="bu">print</span>(<span class="st">"Reading data matrix"</span>)</span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a>  adata <span class="op">=</span> sc.read_10x_mtx(<span class="bu">file</span>)</span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb35-6"><a href="#cb35-6" aria-hidden="true" tabindex="-1"></a>  <span class="bu">print</span>(<span class="st">"Adding gene metadata"</span>)</span>
<span id="cb35-7"><a href="#cb35-7" aria-hidden="true" tabindex="-1"></a>  gene_annot <span class="op">=</span> adata.var.merge(gene_annot, how<span class="op">=</span><span class="st">"left"</span>, on<span class="op">=</span><span class="st">"gene_ids"</span>)</span>
<span id="cb35-8"><a href="#cb35-8" aria-hidden="true" tabindex="-1"></a>  gene_annot <span class="op">=</span> gene_annot.set_index(adata.var.index)</span>
<span id="cb35-9"><a href="#cb35-9" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb35-10"><a href="#cb35-10" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (adata.var[<span class="st">"gene_ids"</span>] <span class="op">==</span> gene_annot[<span class="st">"gene_ids"</span>]).<span class="bu">all</span>():</span>
<span id="cb35-11"><a href="#cb35-11" aria-hidden="true" tabindex="-1"></a>    adata.var <span class="op">=</span> gene_annot</span>
<span id="cb35-12"><a href="#cb35-12" aria-hidden="true" tabindex="-1"></a>  <span class="cf">else</span>: </span>
<span id="cb35-13"><a href="#cb35-13" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="st">"The gene order was not the same!"</span>)</span>
<span id="cb35-14"><a href="#cb35-14" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span></span>
<span id="cb35-15"><a href="#cb35-15" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb35-16"><a href="#cb35-16" aria-hidden="true" tabindex="-1"></a>  vars_to_keep <span class="op">=</span> adata.var[<span class="st">"chrom"</span>].isin([<span class="bu">str</span>(i) <span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">1</span>, <span class="dv">23</span>)] <span class="op">+</span> [<span class="st">"X"</span>, <span class="st">"Y"</span>, <span class="st">"MT"</span>])</span>
<span id="cb35-17"><a href="#cb35-17" aria-hidden="true" tabindex="-1"></a>  adata <span class="op">=</span> adata[:, vars_to_keep].copy()</span>
<span id="cb35-18"><a href="#cb35-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-19"><a href="#cb35-19" aria-hidden="true" tabindex="-1"></a>  adata.var[<span class="st">"mt"</span>] <span class="op">=</span> adata.var[<span class="st">"chrom"</span>] <span class="op">==</span> <span class="st">"MT"</span></span>
<span id="cb35-20"><a href="#cb35-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-21"><a href="#cb35-21" aria-hidden="true" tabindex="-1"></a>  <span class="bu">print</span>(<span class="st">"Calculating QC metrics and filtering"</span>)</span>
<span id="cb35-22"><a href="#cb35-22" aria-hidden="true" tabindex="-1"></a>  sc.pp.calculate_qc_metrics(adata, </span>
<span id="cb35-23"><a href="#cb35-23" aria-hidden="true" tabindex="-1"></a>                             qc_vars<span class="op">=</span>[<span class="st">"mt"</span>], </span>
<span id="cb35-24"><a href="#cb35-24" aria-hidden="true" tabindex="-1"></a>                             inplace<span class="op">=</span><span class="va">True</span>, percent_top<span class="op">=</span>[<span class="dv">20</span>], </span>
<span id="cb35-25"><a href="#cb35-25" aria-hidden="true" tabindex="-1"></a>                             log1p<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb35-26"><a href="#cb35-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-27"><a href="#cb35-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-28"><a href="#cb35-28" aria-hidden="true" tabindex="-1"></a>  adata.obs[<span class="st">"counts_outlier"</span>] <span class="op">=</span> is_outlier(adata, <span class="st">"log1p_total_counts"</span>, <span class="dv">5</span>)</span>
<span id="cb35-29"><a href="#cb35-29" aria-hidden="true" tabindex="-1"></a>  adata.obs[<span class="st">"genes_outlier"</span>] <span class="op">=</span> is_outlier(adata, <span class="st">"log1p_n_genes_by_counts"</span>, <span class="dv">5</span>)</span>
<span id="cb35-30"><a href="#cb35-30" aria-hidden="true" tabindex="-1"></a>  adata.obs[<span class="st">"topgenes_outlier"</span>] <span class="op">=</span> is_outlier(adata, <span class="st">"pct_counts_in_top_20_genes"</span>, <span class="dv">5</span>)</span>
<span id="cb35-31"><a href="#cb35-31" aria-hidden="true" tabindex="-1"></a>  adata.obs[<span class="st">"mito_outlier"</span>] <span class="op">=</span> is_outlier(adata, <span class="st">"pct_counts_mt"</span>, <span class="dv">3</span>) <span class="op">|</span> (adata.obs[<span class="st">"pct_counts_mt"</span>] <span class="op">&gt;</span> <span class="dv">8</span>)</span>
<span id="cb35-32"><a href="#cb35-32" aria-hidden="true" tabindex="-1"></a>  adata.obs[<span class="st">"outlier"</span>] <span class="op">=</span> adata.obs[<span class="st">"genes_outlier"</span>] <span class="op">|</span> adata.obs[<span class="st">"genes_outlier"</span>] <span class="op">|</span> adata.obs[<span class="st">"topgenes_outlier"</span>] <span class="op">|</span> adata.obs[<span class="st">"mito_outlier"</span>]</span>
<span id="cb35-33"><a href="#cb35-33" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb35-34"><a href="#cb35-34" aria-hidden="true" tabindex="-1"></a>  adata <span class="op">=</span> adata[<span class="op">~</span>adata.obs[<span class="st">"outlier"</span>]].copy()</span>
<span id="cb35-35"><a href="#cb35-35" aria-hidden="true" tabindex="-1"></a>  sc.pp.filter_genes(adata, min_cells<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb35-36"><a href="#cb35-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-37"><a href="#cb35-37" aria-hidden="true" tabindex="-1"></a>  <span class="bu">print</span>(<span class="st">"Applying normalisations"</span>)</span>
<span id="cb35-38"><a href="#cb35-38" aria-hidden="true" tabindex="-1"></a>  adata.layers[<span class="st">"counts"</span>] <span class="op">=</span> adata.X.copy()</span>
<span id="cb35-39"><a href="#cb35-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-40"><a href="#cb35-40" aria-hidden="true" tabindex="-1"></a>  sc.pp.normalize_total(adata, target_sum<span class="op">=</span><span class="va">None</span>)</span>
<span id="cb35-41"><a href="#cb35-41" aria-hidden="true" tabindex="-1"></a>  sc.pp.log1p(adata)</span>
<span id="cb35-42"><a href="#cb35-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-43"><a href="#cb35-43" aria-hidden="true" tabindex="-1"></a>  <span class="cf">return</span> adata</span>
<span id="cb35-44"><a href="#cb35-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-45"><a href="#cb35-45" aria-hidden="true" tabindex="-1"></a><span class="co"># our outlier detection function from before</span></span>
<span id="cb35-46"><a href="#cb35-46" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> is_outlier(adata, metric: <span class="bu">str</span>, nmads: <span class="bu">int</span>):</span>
<span id="cb35-47"><a href="#cb35-47" aria-hidden="true" tabindex="-1"></a>  M <span class="op">=</span> adata.obs[metric]</span>
<span id="cb35-48"><a href="#cb35-48" aria-hidden="true" tabindex="-1"></a>  outlier <span class="op">=</span> (M <span class="op">&lt;</span> np.median(M) <span class="op">-</span> nmads <span class="op">*</span> median_abs_deviation(M)) <span class="op">|</span> (</span>
<span id="cb35-49"><a href="#cb35-49" aria-hidden="true" tabindex="-1"></a>      np.median(M) <span class="op">+</span> nmads <span class="op">*</span> median_abs_deviation(M) <span class="op">&lt;</span> M</span>
<span id="cb35-50"><a href="#cb35-50" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb35-51"><a href="#cb35-51" aria-hidden="true" tabindex="-1"></a>  <span class="cf">return</span> outlier</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We apply this function to each of our samples, looping through each of them based on our sample information DataFrame:</p>
<div id="e5a181c5" class="cell" data-execution_count="33">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>sample_info <span class="op">=</span> pd.read_table(<span class="st">"Data/sample_sheet.tsv"</span>)</span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>sample_info <span class="op">=</span> sample_info.rename(columns<span class="op">=</span>{<span class="st">"Sample"</span>: <span class="st">"sample_id"</span>, </span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a>                                          <span class="st">"SampleName"</span>: <span class="st">"sample_name"</span>, </span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a>                                          <span class="st">"SampleGroup"</span>: <span class="st">"sample_group"</span>})</span>
<span id="cb36-5"><a href="#cb36-5" aria-hidden="true" tabindex="-1"></a>sample_info</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="31">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">sample_id</th>
<th data-quarto-table-cell-role="th">sample_name</th>
<th data-quarto-table-cell-role="th">sample_group</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>SRR9264343</td>
<td>ETV6-RUNX1_1</td>
<td>ETV6-RUNX1</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>SRR9264344</td>
<td>ETV6-RUNX1_2</td>
<td>ETV6-RUNX1</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>SRR9264345</td>
<td>ETV6-RUNX1_3</td>
<td>ETV6-RUNX1</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>SRR9264346</td>
<td>ETV6-RUNX1_4</td>
<td>ETV6-RUNX1</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>SRR9264347</td>
<td>HHD_1</td>
<td>HHD</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">5</td>
<td>SRR9264348</td>
<td>HHD_2</td>
<td>HHD</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">6</td>
<td>SRR9264349</td>
<td>PRE-T_1</td>
<td>PRE-T</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">7</td>
<td>SRR9264350</td>
<td>PRE-T_2</td>
<td>PRE-T</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">8</td>
<td>SRR9264351</td>
<td>PBMMC_1</td>
<td>PBMMC</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">9</td>
<td>SRR9264353</td>
<td>PBMMC_2</td>
<td>PBMMC</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">10</td>
<td>SRR9264354</td>
<td>PBMMC_3</td>
<td>PBMMC</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<p>To make it doable computationally, we restrict things to a few samples only:</p>
<div id="40e86bd9" class="cell" data-execution_count="34">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="co"># create a dictionary to hold all AnnData objects</span></span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>adata <span class="op">=</span> <span class="bu">dict</span>.fromkeys(sample_info[<span class="st">"sample_id"</span>])</span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a><span class="co"># read all samples</span></span>
<span id="cb37-5"><a href="#cb37-5" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> sample <span class="kw">in</span> adata.keys():</span>
<span id="cb37-6"><a href="#cb37-6" aria-hidden="true" tabindex="-1"></a>  <span class="bu">print</span>(<span class="st">"Reading </span><span class="sc">{}</span><span class="st">"</span>.<span class="bu">format</span>(sample))</span>
<span id="cb37-7"><a href="#cb37-7" aria-hidden="true" tabindex="-1"></a>  adata[sample] <span class="op">=</span> sc_preprocess(</span>
<span id="cb37-8"><a href="#cb37-8" aria-hidden="true" tabindex="-1"></a>    <span class="st">"Data/CellRanger_Outputs/</span><span class="sc">{}</span><span class="st">/outs/filtered_feature_bc_matrix/"</span>.<span class="bu">format</span>(sample),</span>
<span id="cb37-9"><a href="#cb37-9" aria-hidden="true" tabindex="-1"></a>    h38_genes</span>
<span id="cb37-10"><a href="#cb37-10" aria-hidden="true" tabindex="-1"></a>    )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Reading SRR9264343
Reading data matrix
Adding gene metadata
Calculating QC metrics and filtering
Applying normalisations
Reading SRR9264344
Reading data matrix
Adding gene metadata
Calculating QC metrics and filtering
Applying normalisations
Reading SRR9264345
Reading data matrix
Adding gene metadata
Calculating QC metrics and filtering
Applying normalisations
Reading SRR9264346
Reading data matrix
Adding gene metadata
Calculating QC metrics and filtering
Applying normalisations
Reading SRR9264347
Reading data matrix
Adding gene metadata
Calculating QC metrics and filtering
Applying normalisations
Reading SRR9264348
Reading data matrix
Adding gene metadata
Calculating QC metrics and filtering
Applying normalisations
Reading SRR9264349
Reading data matrix
Adding gene metadata
Calculating QC metrics and filtering
Applying normalisations
Reading SRR9264350
Reading data matrix
Adding gene metadata
Calculating QC metrics and filtering
Applying normalisations
Reading SRR9264351
Reading data matrix
Adding gene metadata
Calculating QC metrics and filtering
Applying normalisations
Reading SRR9264353
Reading data matrix
Adding gene metadata
Calculating QC metrics and filtering
Applying normalisations
Reading SRR9264354
Reading data matrix
Adding gene metadata
Calculating QC metrics and filtering
Applying normalisations</code></pre>
</div>
</div>
<p>Our next step is to combine all these separate objects into a single AnnData object, with information about the sample that each barcode comes from.</p>
<p>Before doing that, we will grab the gene metadata, which unfortunately gets lost when combining multiple AnnData objects (see <a href="https://discourse.scverse.org/t/loosing-anndata-var-layer-when-using-sc-concat/1605">this post</a> for details).</p>
<div id="bd65dc8c" class="cell" data-execution_count="35">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="co"># grab all var DataFrames from our dictionary</span></span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>all_var <span class="op">=</span> [x.var <span class="cf">for</span> x <span class="kw">in</span> adata.values()]</span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a><span class="co"># concatenate them</span></span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a>all_var <span class="op">=</span> pd.concat(all_var, join<span class="op">=</span><span class="st">"outer"</span>)</span>
<span id="cb39-5"><a href="#cb39-5" aria-hidden="true" tabindex="-1"></a><span class="co"># we keep only some columns of interest</span></span>
<span id="cb39-6"><a href="#cb39-6" aria-hidden="true" tabindex="-1"></a>all_var <span class="op">=</span> all_var[[<span class="st">"gene_ids"</span>, <span class="st">"gene_name"</span>, <span class="st">"chrom"</span>]]</span>
<span id="cb39-7"><a href="#cb39-7" aria-hidden="true" tabindex="-1"></a><span class="co"># remove duplicates</span></span>
<span id="cb39-8"><a href="#cb39-8" aria-hidden="true" tabindex="-1"></a>all_var <span class="op">=</span> all_var[<span class="op">~</span>all_var.duplicated()]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now we have all the gene metadata, we proceed with combining all these objects together using scanpy’s concatenation function:</p>
<div id="a5b0970b" class="cell" data-execution_count="36">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>adata <span class="op">=</span> sc.concat(adata, join<span class="op">=</span><span class="st">"outer"</span>, label<span class="op">=</span><span class="st">"sample_id"</span>, index_unique<span class="op">=</span><span class="st">"-"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>After concatenation, we can add back the gene metadata that we collected earlier:</p>
<div id="2d8657e5" class="cell" data-execution_count="37">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>adata.var <span class="op">=</span> all_var.loc[adata.var_names]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>And we add the information about our samples to the cells’ metadata:</p>
<div id="781accfe" class="cell" data-execution_count="38">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>adata.obs <span class="op">=</span> (</span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a>  adata.obs</span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a>  .merge(sample_info, how<span class="op">=</span><span class="st">"left"</span>, on<span class="op">=</span><span class="st">"sample_id"</span>)</span>
<span id="cb42-4"><a href="#cb42-4" aria-hidden="true" tabindex="-1"></a>  .set_axis(adata.obs.index)</span>
<span id="cb42-5"><a href="#cb42-5" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Note we used <code>.set_axis(adata.obs.index)</code> to ensure we retained the rowname indexes of the table as the original (i.e.&nbsp;the barcode names). If we don’t do this, it will generate an invalid anndata object.</p>
<p>Finally, with the full data merged, we save this object for downstream analysis:</p>
<div id="9dbb8ea8" class="cell" data-execution_count="39">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>os.makedirs(<span class="st">"results/scanpy"</span>)</span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a>adata.write(<span class="st">"results/scanpy/caron_filtered_full.hdf5"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We are now ready for our downstream analysis in part II.</p>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      return note.innerHTML;
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>